src/Spectre.libGaussianMixtureModelling/MaximizationRunnerRef.h

```diff
@@ -48,81 +48,114 @@ limitations under the License.
+    {
+        spectrum.mzs.size(); // Needed to avoid warning of unused param.
+    }
+
+    /// <summary>
+    /// Updates weights, means and tandard deviations in a single run.
+    /// </summary>
+    /// <param name="affilationMatrix">Matrix symbolising the probability
+    /// of affilation of each sample to a certain gaussian component.
+    /// </param>
+    /// <param name="spectrum">Spectrum to base calculations on.</param>
     /// <param name="components">Gaussian components to be updated</param>
-    /// <exception cref="ArgumentNullException">Thrown when either of mzArray or intensities pointers are null</exception>
-    MaximizationRunnerRef(SpectrumView spectrum, Matrix<DataType> &affilationMatrix, std::vector<GaussianComponent> &components)
-        : m_Spectrum(spectrum), m_AffilationMatrix(affilationMatrix), m_Components(components)
+    void Maximization(Matrix<DataType> &affilationMatrix,
+        SpectrumView spectrum, GaussianMixtureModel& components)
     {
+        UpdateWeights(affilationMatrix, spectrum, components);
+        UpdateMeans(affilationMatrix, spectrum, components);
+        UpdateStdDeviations(affilationMatrix, spectrum, components);
     }
 
+private:
     /// <summary>
-    /// Updates weights in gaussian components, based on affilation (gamma) matrix.
+    /// Updates weights in gaussian components, based on affilation matrix.
     /// </summary>
-    void UpdateWeights()
+    /// <param name="affilationMatrix">Matrix symbolising the probability
+    /// of affilation of each sample to a certain gaussian component.
+    /// </param>
+    /// <param name="spectrum">Spectrum to base calculations on.</param>
+    /// <param name="components">Gaussian components to be updated</param>
+    void UpdateWeights(Matrix<DataType> &affilationMatrix,
+        SpectrumView spectrum, GaussianMixtureModel& components)
     {
         // This part conducts the instruction:
         // "Using the gamma calculated in the Expectation step, 
         // calculate in the following order, for every k: (phi part)"
         // from the document
         DataType totalDataSize = 0.0;
-        for (unsigned i = 0; i < (unsigned)m_Spectrum.mzs.size(); i++)
+        for (unsigned i = 0; i < (unsigned)spectrum.mzs.size(); i++)
         {
-            totalDataSize += m_Spectrum.intensities[i];
+            totalDataSize += spectrum.intensities[i];
         }
 
-        const unsigned numberOfComponents = (unsigned)m_Components.size();
+        const unsigned numberOfComponents = (unsigned)components.size();
         for (unsigned k = 0; k < numberOfComponents; k++)
         {
             DataType weight = 0.0;
-            for (unsigned i = 0; i < (unsigned)m_Spectrum.mzs.size(); i++)
+            for (unsigned i = 0; i < (unsigned)spectrum.mzs.size(); i++)
             {
-                weight += m_AffilationMatrix[i][k] * m_Spectrum.intensities[i];
+                weight += affilationMatrix[k][i] * spectrum.intensities[i];
```

Comment:
Here is probably great underflow. You are summing like 200 000 floats in real data. Really small floats, coming from gaussian distribution. Check [this thread](https://stackoverflow.com/questions/2148149/how-to-sum-a-large-number-of-float-number).